src/libmongoc/src/mongoc/mongoc-topology-background-monitoring.c

```diff
@@ -0,0 +1,834 @@
+   uint64_t min_heartbeat_frequency_ms;
+   int64_t connect_timeout_ms;
+   bool use_tls;
+#ifdef MONGOC_ENABLE_SSL
+   mongoc_ssl_opt_t *ssl_opts;
+#endif
+   mongoc_uri_t *uri;
+   mongoc_host_list_t host;
+   /* A custom initiator may be set if a user provides overrides to create a
+    * stream. */
+   mongoc_stream_initiator_t initiator;
+   void *initiator_context;
+   mongoc_stream_t *stream;
+   int64_t request_id;
+   mongoc_apm_callbacks_t apm_callbacks;
+   void *apm_context;
+} mongoc_server_monitor_t;
+
+/* Called only from server monitor thread.
+ * Caller must hold no locks (user's callback may lock topology mutex).
+ * Locks APM mutex.
+ */
+static void
+_server_monitor_heartbeat_started (mongoc_server_monitor_t *server_monitor)
+{
+   if (server_monitor->apm_callbacks.server_heartbeat_started) {
+      mongoc_apm_server_heartbeat_started_t event;
+      event.host = &server_monitor->host;
+      event.context = server_monitor->apm_context;
+      bson_mutex_lock (&server_monitor->topology->apm_mutex);
+      server_monitor->apm_callbacks.server_heartbeat_started (&event);
+      bson_mutex_unlock (&server_monitor->topology->apm_mutex);
+   }
+}
+
+/* Called only from server monitor thread.
+ * Caller must hold no locks (user's callback may lock topology mutex).
+ * Locks APM mutex.
+ */
+static void
+_server_monitor_heartbeat_succeeded (mongoc_server_monitor_t *server_monitor,
+                                     const bson_t *reply,
+                                     int64_t duration_usec)
+{
+   if (server_monitor->apm_callbacks.server_heartbeat_succeeded) {
+      mongoc_apm_server_heartbeat_succeeded_t event;
+      event.host = &server_monitor->host;
+      event.context = server_monitor->apm_context;
+      event.reply = reply;
+      event.duration_usec = duration_usec;
+      bson_mutex_lock (&server_monitor->topology->apm_mutex);
+      server_monitor->apm_callbacks.server_heartbeat_succeeded (&event);
+      bson_mutex_unlock (&server_monitor->topology->apm_mutex);
+   }
+}
+
+/* Called only from server monitor thread.
+ * Caller must hold no locks (user's callback may lock topology mutex).
+ * Locks APM mutex.
+ */
+static void
+_server_monitor_heartbeat_failed (mongoc_server_monitor_t *server_monitor,
+                                  const bson_error_t *error,
+                                  int64_t duration_usec)
+{
+   if (server_monitor->apm_callbacks.server_heartbeat_failed) {
+      mongoc_apm_server_heartbeat_failed_t event;
+      event.host = &server_monitor->host;
+      event.context = server_monitor->apm_context;
+      event.error = error;
+      event.duration_usec = duration_usec;
+      bson_mutex_lock (&server_monitor->topology->apm_mutex);
+      server_monitor->apm_callbacks.server_heartbeat_failed (&event);
+      bson_mutex_unlock (&server_monitor->topology->apm_mutex);
+   }
+}
+
+static bool
+_server_monitor_cmd_send (mongoc_server_monitor_t *server_monitor,
```

Comment:
This is closely modelled from mongoc-async-cmd.c, and will be modified to support streamable ismaster (will need to send OP_MSG exhaust cursors if the server monitor transitions to streaming).